contributors/design-proposals/api-machinery/udp-portforwarding.md

```diff
@@ -0,0 +1,109 @@
+# UDP Port Forwarding in Kubernetes
+
+Author: @shiywang
+
+## Abstract
+
+This describes an approach for providing support for:
+
+- port forwarding to containers with UDP protocol
+
+## Background && Motivation
+
+Fixes user request of https://github.com/kubernetes/kubernetes/issues/47862
+
+
+## Constraints and Assumptions
+
+- We already have an implemention with SPDY for TCP, the UDP implement basically reuse the SPDY framework, so it would be UDP through SPDY, the change here is remove the stream copy since UDP doesn't not has the same stream attribute as SPDY/TCP, so UDP will use datagram and `write` and `send` to communicate, socat will use unix domain socket as a middle layer to cross the boundaries between pod and Kubelet. and since TCP/SPDY doesn't have a message boundaries but UDP does, we also need to design an simple protocol for how to divide message read from SPDY back and forth.
```

Comment:
Our goal with port-forward/exec is to remove SPDY support across the board and replace it with web sockets.  However, the websockets implementation for port-forward is not scalable (it cannot multiplex multiple streams like SPDY) and therefore cannot be used.  I don't want to add anything to exec/port-forward until we fix the websocket port forward issue by implementing a new multiple stream sub-protocol and deprecate SPDY.

I don't want to add anything new to SPDY.

Reply:
Hi @smarterclayton I am also interested in extending the port-forwarding and other streaming functionalities. What's the status of websocket multiplexing sub-protocol? Any blockers or anything we can help with this effort?
I noticed that there is already minimum multiple stream sub-protocol support in current implementation of api-server and kubelet, but the number of streams, aka channels, is negotiated during handshaking and a stream is identified by the first byte of a frame. Am I correct?